File: LinkedList/reverseKGroup.js
//25. K 个一组翻转链表
function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
}
const ListEntity = require("./makeListFromArray");
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function(head, k) {
    let result = null;
    let cur = null;
    while(head!==null){
        let [rlist,end] = reverseList(head,k);
        if(result === null){
            result = rlist;
            cur = result;
        }else{
            cur.next = rlist;
        }
        while(cur.next){
            cur = cur.next;
        }
        head = end;
    }
    console.log(ListEntity.getArrayFromList(result));
    function reverseList(start,len){//翻转
        let tail = null;
        let shead = null;
        let set = new Set;
        set.add(start);
        let count = 0;
        while(count++<len){
            if(start === null) return [Array.from(set)[0],null];
            shead = new ListNode(start.val,null);
            if(tail){
                shead.next = tail;
            }
            tail = shead;
            start = start.next;
        }
        console.log(shead);
        return [shead,start];
    };
    return result;
};
let l1 = ListEntity.makeListFromArray([1,2,3,4,5]);
reverseKGroup(l1,2);

File: C version/link/reverseKGroup.c
#include <stdio.h>
#include <stdlib.h>
#include "tools.h"


struct ListNode* reverseKGroup(struct ListNode* head, int k){
    if(k==1) return head;
    struct ListNode* pro = head;
    struct ListNode* start = NULL;
    struct ListNode* end = NULL;
    struct ListNode* sHead = NULL;
    int count = 0;
    struct ListNode* zip = NULL;
    while(head!=NULL){
        if(count%k==0){
            printf("%d\n",head->val);
            if(end!=NULL){
                printf("[%d %d]\n",end->val,head->val);
                if(zip!=NULL){
                    zip->next = end;
                }
                zip = pro;
                if(sHead==NULL) sHead = end;
            } 
            pro = head;
            start = NULL;
            end = NULL;
        }
        start = head;
        head = head->next;
        start->next = end;
        end = start;
        count++;
    }
    struct ListNode* prev = NULL;
    struct ListNode* next = end;
    while(end!=NULL){
        end = end->next;
        next->next = prev;
        prev = next;
        next = end;
    }
    if(zip!=NULL) zip->next = prev;
    while(sHead!=NULL){
        printf("%d\n",sHead->val);
        sHead = sHead->next;
    }
    return sHead;
}
void main(){
    int arr[5] = {1,2};
    struct ListNode*l1 = makeLinkFromList(arr,2);
    reverseKGroup(l1,2);
}